### PR TITLE
Smileys refactor

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

File: smileys-app/schema.psql [hunk 1]
```diff
@@ -1,12 +1,13 @@
 CREATE SCHEMA smileys;
 
 CREATE TABLE smileys.trail (
-  id SERIAL PRIMARY KEY,
   entries JSON NOT NULL,
   username TEXT NOT NULL,
   smiley SMALLINT NOT NULL,
   day DATE NOT NULL,
-  constraint one_a_day UNIQUE (username, day)
+  PRIMARY KEY (username, day)  -- this is natural primary key
```
Comment (mixman) [inline smileys-app/schema.psql]: What's the benefit of a natural PK here?

File: smileys-app/schema.psql [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 );
 
-CREATE INDEX ON smileys.trail (day);
+-- primary key creates an index,
+-- but let's have this way around too.
+CREATE INDEX ON smileys.trail (day, username);
```
Comment (mixman) [inline smileys-app/schema.psql]: Does this also allow efficient lookups on day alone (start-date, end-date -range lookups)?

File: smileys-app/src/Futurice/App/Smileys/Logic.hs [hunk 3]
```diff
@@ -14,43 +14,43 @@ import qualified FUM
...
-    let today = utctDay now
-    withResource (ctxPostgresPool ctx) $ \conn -> do
-        res <- q conn (fromMaybe today start) (fromMaybe today end)
-        return $ res
-        where
-          q :: (MonadIO m, MonadBaseControl IO m) => Postgres.Connection -> Day -> Day -> m [Smileys]
-          q conn s e = liftIO $ Postgres.query conn
-                "SELECT entries, username, smiley, day FROM smileys.trail WHERE day >= ? AND day <= ?"
-                (s, e)
+getSmileys ctx mfum start end =
```
Comment (mixman) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: I'll move this to a getSmileysForUser as getSmileys is needed also as-is.

File: smileys-app/schema.psql (hunk 2 above)
Comment (phadej) [inline smileys-app/schema.psql]: yes.

File: smileys-app/src/Futurice/App/Smileys/Logic.hs (hunk 3 above)
Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: I'd rather have three
- `getSmileys`
- `getSmileysForAll` (without SSO User)
- `getSmileysForUser` (Where FUM.UserName` is a `Capture`

Comment (phadej) [inline smileys-app/src/Futurice/App/Smileys/Logic.hs]: Or last two can be the same, if `FUM.UserName` is a `QueryParam` (return all if `Nothing`).

File: smileys-app/schema.psql (hunk 2 above)
Comment (phadej) [inline smileys-app/schema.psql]: > A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns. 

etc. https://www.postgresql.org/docs/9.6/static/indexes-multicolumn.html

File: smileys-app/schema.psql (hunk 1 above)
Comment (phadej) [inline smileys-app/schema.psql]: No need for the artificial one.

Comment (mixman) [inline smileys-app/schema.psql]: They are two ways of achieving the same goal; both have pro/cons depending on the situation. Was wondering why in this case a NPK.

**Comment by phadej** [comment]
travis failure was the timeout.

**Pull request merged**